[docs] Explain the builtin commands and the separating script for script run stage (#5127)
pipe-cd/pipecd: docs/content/en/docs-dev/user-guide/managing-application/customizing-deployment/script-run.md

![](/images/script-run.png)

### Execute the script file

If your script is so long, you can separate the script as a file.
You can put the file with the app.pipecd.yaml in the same dir and then you can execute the script like this.

```
apiVersion: pipecd.dev/v1beta1
kind: KubernetesApp
spec:
  name: script-run
  labels:
    env: example

>     team: product

Reviewer: [nits] I think `labels` section is not necessary for the explanation.

  pipeline:
    stages:
      - name: SCRIPT_RUN
        with:
          run: |
            sh script.sh
```

```
.
├── app.pipecd.yaml
└── script.sh
```

> ## Builtin command

Reviewer: [nits] 
```suggestion
## Builtin commands
```

Currently, you can use the commands which are installed in the environment for the piped.

For example, If you are using the container platform and the offcial piped container image, you can use the command below.
- git
- ssh
- jq
- curl

> - and the builtin commands installed in the base image.

Reviewer 2: Users can also use piped installed tools such as terraform, kubectl, helm, kustomize, etc 👀

Author: @khanhtc1202 
Thank you for the comment.
I think it might be unexpected usage for these tools.
WDYT?

Reviewer 2: Sorry, I don't really get your point. Could you explain it :pray:

Author: 📝 Discussed about it in person. I agree that users can use the commands installed for the piped.
- we can use them on the script run stage
- but this may change in the future

Author: fixed on [48d7819](https://github.com/pipe-cd/pipecd/pull/5127/commits/48d781964683100dcd86cc0f1dfecde6c2a572a7)

Reviewer 3: The tool binaries in `$PIPED_TOOLS_DIR` are installed just before it's needed.
So, the binaries are sometimes absent, such as when the SCRIPT_RUN stage is the first stage of the pipeline and the binaries are not installed yet.

Reviewer 2: I understand your both point now, and agree with @Warashi If it isn't always available, the it should not be listed. But, it could be better to mention that some tools could be provided under PIPED_TOOLS_DIR so that users can check it out by themself, wdyt? @ffjlabo @Warashi

Reviewer 3: I agree with @khanhtc1202 .
It's okay to mention that some tools may be available under PIPED_TOOLS_DIR.
But I think we also have to mention that the tools can be absent sometimes.

Author: @khanhtc1202 @Warashi 
> So, the binaries are sometimes absent, such as when the SCRIPT_RUN stage is the first stage of the pipeline and the binaries are not installed yet.

I forgot it. So I think we should not explain it.

It might be confusing for users who use SCRIPT_RUN that the command does not exist until the first deployment is executed.
If the piped administrator and the app administrator are different, I don't think there is a way for the app administrator to check the dir.

WDYT?

Reviewer 3: The app administrator can use the SCRIPT_RUN stage, such as executing `ls`, to check whether the commands exist.
But I think it's nonsense to check whether the commands exist. If the administrator has to use a command, they should install it in the SCRIPT_RUN stage or ask the piped admin to use a custom container image containing the command.

I think it's okay to mention some tools are sometimes available under PIPED_TOOLS_DIR because it's a fact.
On the other hand, users must use those tools carefully, and we must explain why users must pay attention.

Author: I tried adding it anyway. [bfb1765](https://github.com/pipe-cd/pipecd/pull/5127/commits/bfb176524c3e29e96c5c11ef6c870479ede2646e)

Author: Plz check it 🙏 @Warashi @khanhtc1202

> The public piped image available in PipeCD main repo (ref: [Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/cmd/piped/Dockerfile)) is based on [alpine](https://hub.docker.com/_/alpine/) and only has a few UNIX command available (ref: [piped-base Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/tool/piped-base/Dockerfile)). 

Reviewer: [nits]

```suggestion
The public piped image available in PipeCD main repo (ref: [Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/cmd/piped/Dockerfile)) is based on [alpine](https://hub.docker.com/_/alpine/) and only has a few UNIX commands available (ref: [piped-base Dockerfile](https://github.com/pipe-cd/pipecd/blob/master/tool/piped-base/Dockerfile)). 
```

> If you want to use your commands, you can:

Reviewer: [nits] I wanna clarify whether the options are isolated or combined.

e.g. 
```suggestion
If you want to use your commands, follow these steps:
```

![](/images/script-run.png)

### Execute the script file

If your script is so long, you can separate the script as a file.
You can put the file with the app.pipecd.yaml in the same dir and then you can execute the script like this.

```
apiVersion: pipecd.dev/v1beta1
kind: KubernetesApp
spec:
  name: script-run
  pipeline:
    stages:
      - name: SCRIPT_RUN
        with:
          run: |
            sh script.sh
```

```
.
├── app.pipecd.yaml
└── script.sh
```

## Builtin commands

Currently, you can use the commands which are installed in the environment for the piped.

For example, If you are using the container platform and the offcial piped container image, you can use the command below.

> - git
> - ssh
> - jq
> - curl
> - kubectl, helm, kustomize, terraform stored in $PIPED_TOOL_DIR 
>   - Be careful. These tools are sometimes absent because they are installed when the first deployment has been done using each tool. Please check $PIPED_TOOL_DIR before you use them.
>   - Binaries can be specified as "kubectl-${version}" or "kubectl". 
>   - If no version is specified on the piped config or app.pipecd.yaml, both the ones with and without default version suffix will be installed. 
>   - Please check the [code](https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/toolregistry/install.go#L28C1-L33C2) about the default versions.
> - and the builtin commands installed in the base image.

Reviewer 2: ```suggestion
- git
- ssh
- jq
- curl
- commands installed by piped in $PIPED_TOOL_DIR (check at runtime)
- built-in commands installed in the base image.
```

How about this? @ffjlabo

Author: @khanhtc1202 
I think we also have to mention that the tools can be absent sometimes.
WDYT?

Reviewer 2: That is why I added the note `(check at runtime)`. I think it would be better not to confuse users by showing too much explanation 🤔

Author: OK, I agree with your point.
I will fix it👌

Author: fixed [48022da](https://github.com/pipe-cd/pipecd/pull/5127/commits/48022daafc84b0beb3e27cb973d4f9e6979a742d)
